Fix splitting image name and image tag on container page

jhadvig opened the pull request:
Since container image could have several formats, eg:
```
busybox

openshift/origin-docker-registry:v3.9.0

quay.io/redhat/origin-web-console-smoke-test@sha256:a27e2ff4b56a53a3d154b6d19d5cf0d189448c575a60ccbd2e2691a549baa20b

brew-pulp-docker11.redhat.address.com:1111/openshift3/ose-console:rhaos-3.11-rhel-8-containers-candidate-11111-222222222
```

we can not split it just based on semicolon. `getContainerImageAndTag` handles all the cases and returns image name and tag.

@spadgett  PTAL

benjaminapetersen commented on frontend/public/components/container.jsx:
```diff
@@ -108,6 +108,21 @@ const Env = ({env}) => {
+  const imageNamePortion = _.trim(semiColonSplit.slice(-2, -1), '@sha256');
+  const imageName = imageNamePortion.split('/').slice(-2).join('/');
+  return [imageName, imageTag];
+};
```
Contemplating an object `{ image: '', tag: '' }` vs array/tuple, but otherwise looks good to me!

spadgett commented on frontend/public/components/container.jsx:
```diff
@@ -108,6 +108,21 @@ const Env = ({env}) => {
+  const imageTag = _.last(semiColonSplit);
+  const imageNamePortion = _.trim(semiColonSplit.slice(-2, -1), '@sha256');
+  const imageName = imageNamePortion.split('/').slice(-2).join('/');
+  return [imageName, imageTag];
```
Suggest we have special handling for the `@sha256` case then use `image.lastIndexOf(':')` in other cases to get the tag.

Would be good to have examples and their output there. I'm not sure exactly how split/slice/join is changing the string.

spadgett replied:
I know the previous code had an array since it used split, but I'd prefer to return an object here with named properties like 

```js
return { imageName, imageTag };
```

jhadvig commented:
@benjaminapetersen @spadgett I've updated the PR and returning an Object and added the comments on how the strings are mutated.

regarding:
>Suggest we have special handling for the @sha256 case then use image.lastIndexOf(':') in other cases to get the tag.

Not really sure whats would be the benefit since this can handle all the container image formats.
PTAL

spadgett commented:
> Not really sure whats would be the benefit since this can handle all the container image formats.

Are we sure? I don't want to assume too much about the image string format.

Splitting and recombining the string seems more complicated and error prone than just using `lastIndexOf`.

```jsx
const index = image.lastIndexOf(':');
if ( index === -1 || /@sha256:/.test(image)) {
  return { imageName: image, imageTag: null };
}
const imageName = image.substr(0, index);
const imageTag = image.substr(index + 1);
return { imageName, imageTag };
```

jhadvig commented:
> Are we sure? I don't want to assume too much about the image string format.

Tested the function with all the container images I could get my hands on and worked with them. But can't tell if that's enough.

Also do we wanna show the image name and tag in the container image points to a DOCKER REF ? Cause the suggested logic will just return the DOCKER REF. Also in case of `busybox` it will just return the name without tag, which should be `latest`.

Asking since in the origin-web-console we have been showing image name and version/tag for the  DOCKER REF. eg.:
```
172.30.1.1:5000/myproject/ruby-ex@sha256:43b9fd7220190bd48bf571525a0dedee7b2b40887bb36920ea0af0289c047fcd
```
we display as:
![download](https://user-images.githubusercontent.com/1668218/42832967-2a5e7d24-89f3-11e8-8e97-82a8a310c7a3.png)

spadgett commented:
Yeah, latest is used if no tag is specified, but I'd rather leave tag as `-` in that case so it's clear it's not explicitly set to latest

> Cause the suggested logic will just return the DOCKER REF.

I think that's good for now. We should consider using the Overflow component for display. We've had complaints that we shorten / reformat the sha. Sometimes the user needs the full value

spadgett commented on frontend/public/components/container.jsx:
```diff
@@ -108,6 +108,20 @@ const Env = ({env}) => {
 
+// Split image string into the Docker image string and tag.
+const getContainerImageAndTag = image => {
+  if (_.isEmpty(image)) {
```
Since image is a string

```js
if (!image)
```

spadgett commented on frontend/public/components/container.jsx:
```diff
@@ -108,6 +108,20 @@ const Env = ({env}) => {
 };
 
+// Split image string into the Docker image string and tag.
+const getContainerImageAndTag = image => {
```
nit: I would call this something like `splitImageString` or `getImageNameAndTag`

spadgett commented on frontend/public/components/container.jsx:
```diff
@@ -118,9 +132,7 @@ const Details = (props) => {
-
-  // Split image string into the Docker image string and tag (aka version) portions. The tag defaults to 'latest'.
-  const [containerImage, containerTag] = container.image ? container.image.split(':') : [null, 'latest'];
+  const containerImage = getContainerImageAndTag(container.image);
```
I think this reads better.

```
const { imageName, imageTag } = getContainerImageAndTag(container.image);
```

spadgett commented on frontend/public/components/container.jsx:
```diff
@@ -108,6 +108,20 @@ const Env = ({env}) => {
   </table>;
 };
 
+// Split image string into the Docker image string and tag.
```
Split image string into the image name and tag.

jhadvig commented:
@spadgett comments addressed
